### PR TITLE
[CICD] Add Ascend NPU unit test support

### PR DESCRIPTION
**Pull request by AlexMa616**
Summary

This PR adds Ascend NPU Unit CI support for TransformerEngine-FL through torch_npu and the FlagOS backend. It extends the existing CI workflow to execute real TE workloads and selected shared PyTorch tests on Ascend 910C.

Changes

Ascend NPU testing

Add real NPU coverage for:

Linear forward and backward
LayerNorm, RMSNorm, and LayerNormLinear
LayerNormMLP
GEMM, softmax, and multi-tensor operations
Unfused Dot Product Attention and MultiheadAttention
TransformerLayer debug and ONNX export paths
Reuse selected portable sanity and numerics tests from the existing PyTorch suites.

Distributed testing

Add support for:

Two-process HCCL execution
TE Linear gradient synchronization
Context Parallel utility tests
Initial non-FP8 distributed numerical validation
CI and coverage

Add Ascend-specific Unit test entry points.
Add raw and aggregated coverage collection.
Fail explicitly when torch_npu or flag_gems is unavailable.
Keep unsupported CUDA-specific features explicitly excluded.
Testing

Verified on Ascend 910C:

PyTorch Unit test execution passed.
PyTorch Debug passed.
PyTorch ONNX passed.
PyTorch Distributed is under validation.
The current PyTorch Unit job failure occurred during coverage artifact upload after the test execution had passed.

Limitations

CUDA Graphs, Flash/Fused Attention, FP8, MXFP8, NVFP4, block scaling, TensorRT integration, and Integration tests are not included in the current Ascend Unit scope.

### CODE REVIEW

File: .github/configs/ascend.yml [hunk 1]
```diff
@@ -1,15 +1,52 @@
...
-labels:
-  - npu
+# Huawei Ascend NPU configuration for TransformerEngine-FL
+# This file follows the same schema as cuda.yml and metax.yml.
+
+hardware_name: ascend
+display_name: 'Huawei Ascend NPU'
+
+# CI image for the Ascend environment
+ci_image: harbor.baai.ac.cn/transformerengine-fl/flagscale:te-fl-triton311
```
Comment (Darryl233) [inline .github/configs/ascend.yml]: Rename the ci_image in accordance with the FlagCICD container image solution

File: .github/workflows/unit_tests_common.yml [hunk 2]
```diff
@@ -108,7 +108,15 @@ jobs:
         uses: actions/checkout@v4
         with:
           fetch-depth: 0
-      
+
+      # Ascend installs the Python/plugin layer and does not need CUDA submodules
+      - name: Checkout Source Code on Ascend
+        if: inputs.platform == 'ascend'
```
Comment (Darryl233) [inline .github/workflows/unit_tests_common.yml]: Check if you can merge these three Checkout Source Code steps into one. Pulling source code shouldn't be a platform-related operation

File: .github/workflows/unit_tests_common.yml [hunk 3]
```diff
@@ -155,42 +181,80 @@ jobs:
           # Ensure log directory exists regardless of volume mount state
           mkdir -p /logs
 
-          # Coverage setup: install once + configure collection via PYTEST_ADDOPTS
+          # Coverage setup: Ascend unit jobs must always produce raw coverage.
           COVERAGE_ENABLED=false
-          if pip3 install coverage pytest-cov --quiet 2>/dev/null; then
+          if [ "${{ inputs.platform }}" = "ascend" ]; then
```
Comment (Darryl233) [inline .github/workflows/unit_tests_common.yml]: Better preinstall dependencies in the image and reduce platform-related implementation in the workflow

File: .github/configs/ascend.yml (hunk 1 above)
Comment (AlexMa616) [reply .github/configs/ascend.yml]: updated

File: .github/workflows/unit_tests_common.yml (hunk 3 above)
Comment (AlexMa616) [reply .github/workflows/unit_tests_common.yml]: updated

File: .github/scripts/setup_cuda.sh [hunk 4]
```diff
@@ -21,5 +26,17 @@ pip install . -v --no-deps --no-build-isolation
 
 echo "===== Step 3: Verify Installation ====="
 python3 tests/pytorch/test_sanity_import.py
+python3 - <<'PY'
+import importlib
+
+tex = importlib.import_module("transformer_engine_torch")
+required = ["multi_tensor_scale", "multi_tensor_compute_scale_and_scale_inv"]
```
Comment (Darryl233) [inline .github/scripts/setup_cuda.sh]: why these two operators are required？

File: .github/scripts/setup_metax.sh [hunk 5]
```diff
@@ -45,6 +48,11 @@ import transformer_engine
 import transformer_engine_torch as te
 print("transformer_engine:", transformer_engine)
 print("transformer_engine_torch:", te)
+required = ["multi_tensor_scale", "multi_tensor_compute_scale_and_scale_inv"]
```
Comment (Darryl233) [inline .github/scripts/setup_metax.sh]: why these two operators are required？

File: .github/scripts/setup_cuda.sh (hunk 4 above)
Comment (AlexMa616) [reply .github/scripts/setup_cuda.sh]: updated

File: .github/scripts/setup_metax.sh (hunk 5 above)
Comment (AlexMa616) [reply .github/scripts/setup_metax.sh]: updated

File: tests/pytorch/test_onnx_export.py [hunk 6]
```diff
@@ -929,7 +953,8 @@ def _test_export_multihead_attention(
         attention_type=attention_type,
         fuse_qkv_params=fuse_qkv_params,
         return_bias=True,
-    ).to(device="cuda")
+        device=te_device_type(),
```
Comment (Darryl233) [inline tests/pytorch/test_onnx_export.py]: device=te_device_type() is not necessary when you have .to(te_device_type())?

Comment (AlexMa616) [reply tests/pytorch/test_onnx_export.py]: updated

File: .github/workflows/integration_tests_common.yml [hunk 7]
```diff
@@ -41,94 +43,72 @@ jobs:
...
         with:
           fetch-depth: 0
-          submodules: recursive
+          submodules: ${{ inputs.checkout_submodules }}
           set-safe-directory: true
 
-      # Metax requires to clean vscode-remote-container
-      - name: Configure Clean Git Env on Metax
-        if: inputs.platform == 'metax'
+      - name: Export Build Environment
```
Comment (Darryl233) [inline .github/workflows/integration_tests_common.yml]: Can we move environment variables into setup_xxx.py script

File: .github/workflows/integration_tests_common.yml [hunk 8]
```diff
@@ -41,94 +43,72 @@ jobs:
...
-        if: inputs.platform == 'metax'
-        uses: actions/checkout@v4
-        with:
-          fetch-depth: 0
-      
       - name: Environment Setup
         if: inputs.setup_script != ''
+        run: bash "$GITHUB_WORKSPACE/${{ inputs.setup_script }}"
+
+      - name: Export Test Environment
```
Comment (Darryl233) [inline .github/workflows/integration_tests_common.yml]: Can we move environment variables into test script?

File: .github/workflows/unit_tests_common.yml [hunk 9]
```diff
@@ -41,168 +70,142 @@ jobs:
...
         with:
           fetch-depth: 0
-          submodules: recursive
+          submodules: ${{ inputs.checkout_submodules }}
           set-safe-directory: true
 
-      # Metax requires to clean vscode-remote-container
-      - name: Configure Clean Git Env on Metax
-        if: inputs.platform == 'metax'
+      - name: Export build environment
```
Comment (Darryl233) [inline .github/workflows/unit_tests_common.yml]: same problem

File: tests/pytorch/debug/test_log.py [hunk 10]
```diff
@@ -68,14 +69,11 @@
 for r in recipes:
     for stat in bare_stats:
         for columnwise_postfix in ["", "_columnwise"]:
-            if (
-                r in ["fp8_current_scaling", "fp8_block_scaling"]
-                and torch.cuda.get_device_capability()[0] < 9
-            ):
-                # hopper is needed for current-scaling, block-scaling
+            if r == "fp8_current_scaling" and not fp8_available:
```
Comment (Darryl233) [inline tests/pytorch/debug/test_log.py]: exlain?

File: .github/workflows/unit_tests_common.yml (hunk 9 above)
Comment (AlexMa616) [reply .github/workflows/unit_tests_common.yml]: updated

File: .github/workflows/integration_tests_common.yml (hunk 8 above)
Comment (AlexMa616) [reply .github/workflows/integration_tests_common.yml]: updated

File: .github/workflows/integration_tests_common.yml (hunk 7 above)
Comment (AlexMa616) [reply .github/workflows/integration_tests_common.yml]: updated

File: tests/pytorch/distributed/run_numerics.py [hunk 11]
```diff
@@ -125,15 +136,18 @@ def main(argv=None, namespace=None):
...
-    test_dict = [
-        test_quantizer,
-        test_quantized_all_gather,
-        test_linear,
-        test_layernorm,
-        test_layernorm_linear,
-        test_layernorm_mlp,
-        test_transformer_layer,
-    ]
+    if bool(int(os.getenv("NVTE_ASCEND_DISTRIBUTED_NUMERICS_SUBSET", "0"))):
```
Comment (Darryl233) [inline tests/pytorch/distributed/run_numerics.py]: what is this?

File: tests/pytorch/debug/test_log.py (hunk 10 above)
Comment (canghaiX) [reply tests/pytorch/debug/test_log.py]: `fp8_current_scaling` only requires general FP8 availability, which is covered by `is_fp8_available()`. `fp8_block_scaling` has stricter requirements, so it uses `is_fp8_block_scaling_available()`.

File: tests/pytorch/distributed/run_numerics.py (hunk 11 above)
Comment (canghaiX) [reply tests/pytorch/distributed/run_numerics.py]: This flag is used by Ascend CI to run a reduced distributed numerics smoke subset. The full numerics suite still contains CUDA-only distributed paths that are not currently supported on Ascend, so this keeps the Ascend job focused on the supported distributed linear coverage.

File: tests/plugin/backend/npu/test_gemm.py
Comment (Darryl233) [inline tests/plugin/backend/npu/test_gemm.py:1]: delete

File: tests/plugin/backend/npu/test_modules.py
Comment (Darryl233) [inline tests/plugin/backend/npu/test_modules.py:1]: delete

File: tests/plugin/backend/npu/npu_patch.py
Comment (Darryl233) [inline tests/plugin/backend/npu/npu_patch.py:1]: delete

File: tests/plugin/backend/npu/run_pytest.py
Comment (Darryl233) [inline tests/plugin/backend/npu/run_pytest.py:1]: delete

File: tests/test_utils/run_ci_test_group.py
Comment (Darryl233) [inline tests/test_utils/run_ci_test_group.py:1]: Delete it if it's useless.

File: tests/plugin/backend/npu/test_gemm.py
Comment (AlexMa616) [reply tests/plugin/backend/npu/test_gemm.py:1]: updtad

File: tests/plugin/backend/npu/test_modules.py
Comment (AlexMa616) [reply tests/plugin/backend/npu/test_modules.py:1]: updated

File: tests/test_utils/run_ci_test_group.py
Comment (AlexMa616) [reply tests/test_utils/run_ci_test_group.py:1]: This is a platform-agnostic test group scheduler used by `unit_tests_common.yml` to execute configured script or pytest groups and report test results.

File: tests/plugin/backend/npu/npu_patch.py
Comment (AlexMa616) [reply tests/plugin/backend/npu/npu_patch.py:1]: Try removing the logic.

File: tests/plugin/backend/npu/run_pytest.py
Comment (AlexMa616) [reply tests/plugin/backend/npu/run_pytest.py:1]: updated

File: tests/pytorch/distributed/run_numerics.py [hunk 12]
```diff
@@ -11,6 +11,17 @@
...
+if os.environ.get("PLATFORM") == "ascend":
+    sys.path.insert(
+        0,
+        os.path.abspath(
+            os.path.join(os.path.dirname(__file__), "../..", "plugin", "backend", "npu")
+        ),
+    )
+    from npu_patch import apply_ascend_npu_patch
+
+    apply_ascend_npu_patch()
```
Comment (Darryl233) [inline tests/pytorch/distributed/run_numerics.py:23]: Check if it's necessary

Comment (canghaiX) [reply tests/pytorch/distributed/run_numerics.py:23]: If this code is removed, the subsequent subprocesses launched for distributed execution will not be able to access the NPU runtime correctly. As a result, the distributed numerics subprocesses may fail during startup validation and subsequent NPU-related calls.

Comment (AlexMa616) [reply tests/pytorch/distributed/run_numerics.py:23]: updated

**Review by Darryl233 — APPROVED** [review]
LGTM

**Review by lxd-cumt — APPROVED** [review]
LGTM
